### DISCUSSION TITLE
Web Api instead of GraphQL

### DISCUSSION THREAD

**Discussion by ricocsharp** [OrchardCMS/OrchardCore#12536] [Q&A]
Hi guys, it is possible in Headless recipe to use a Rest Web Api instead of GraphQl  as the communications interface with client apis? if it is so is there is any example available.

Thnks in advance

**Comment by fuzl-llc** [comment]
I had this question a few weeks ago and wasn't able to find examples in the sense of specific tutorials, but it is definitely possible. Remember Orchard is *just* and ASP.NET app with references to Orchard DLLs. Therefore, you can do anything you can do in an ASP.NET site in an Orchard Core site. I created a regular old controller like I've done in other ASP.NET Core Web API sites I've done. The main difference is you have to use things like IContentManager and IAuthorizationService to implement your logic to get and return data.

And actually, in looking at my code, I see a big note at the top saying my code was based on OrchardCore's /src/OrchardCore.Modules.Cms/OrchardCore.Contents/Controllers/ApiController.cs class. Check it out and give it a try.

**Reply by Skrypt** [reply]
There is a contrib module to add Swagger UI here for everything related with REST API's in OC.

https://github.com/ThisNetWorks/ThisNetWorks.OrchardCore.OpenApi

**Reply by ricocsharp** [reply]
This contrib does allows to generate client-side code?

**Comment by fuzl-llc** [comment]
And again in the spirit of "OC is *just* an ASP.NET app" you can also use swagger directly without the above by referencing the NSwag.AspNetCore nuget package. I tried the above originally and it is great but in my case I'm using RTK query which has its own capability to generate client-side code from the exposed API... and I wanted to expose my own custom types from the API rather than the OC types directly (i.e. I manually mapped OC types server-side to the client-side objects I wanted). It is more work than using the above but you end up with an exposed Open API specification you have a little more control over, e.g. you can't tell I'm using Orchard from my swagger doc because my DTOs are simplified.

**Reply by Skrypt** [reply]
This Contrib project generates also DTO's from OC Content Types for simplification of use. The reason why it is not included in OC is that this kind of module will be generally customized per need. As you did. 😉
